Add sql vs typeql page (#1005)
graknlabs/docs: core-concepts/modules/ROOT/pages/typeql/sql-vs-typeql.adoc

Reviewer: Overall notes:
- I thought we were using sentence case in section titles? Should be fixed if so.
- Placing one sentence per line makes suggestions in a PR easier, and git diffs more meaningful. I suggest reformatting with that in mind.

Author: Fixed point 1 - point 2 isn't something i think we have an agreed upon style with! I agree with reviewing case, but I do think it's more awkward than being able to write a full paragraph in docs!

= SQL versus TypeQL guide
:test-typeql: linear

This guide helps SQL users understand TypeQL by mapping familiar SQL constructs to their TypeQL equivalents.
TypeQL is a declarative query language for TypeDB that uses pattern matching rather than table joins.

[NOTE]
====
TypeQL operates on a graph data model (entities, relations, and attributes) rather than tables and rows.
This fundamental difference means that some SQL concepts map directly, while others require a different approach.
====

== Schema Setup

For the examples in this guide, we'll use a simple schema:

[,typeql]
----
#!test[schema]
define
  entity person, owns name, owns email, owns age;
  attribute name, value string;
  attribute email, value string;
  attribute age, value integer;
  relation employment, relates employee, relates employer;
  person plays employment:employee;
  entity company, owns name, plays employment:employer;
  company owns name;
#!test[write, commit]

> insert

Reviewer: Is this legal?

  $alice isa person, has name "Alice", has email "alice@example.com", has age 30;
  $bob isa person, has name "Bob", has email "bob@example.com", has age 25;
  $charlie isa person, has name "Charlie", has email "charlie@example.com", has age 35;
  $acme isa company, has name "Acme Corp";
  $e1 isa employment, links (employee: $alice, employer: $acme);
----
====

> == Basic Query Operations

Reviewer: Perhaps "Data Retrieval" to better contrast with "Data Modification" and "Aggregation" later on?

=== SELECT → match + fetch/select

In SQL, `SELECT` retrieves data from tables. In TypeQL, `match` finds data by pattern, and `fetch` or `select` projects the results.

**SQL:**

> [source,sql]

Reviewer: Minor comment: using `[source,sql]` and `[,typeql]` is inconsistent. Just `[,sql]` is fine.

----
SELECT name, email FROM person;
----

**TypeQL:**

> In TypeQL, use SELECT to filter out unwanted data:

Reviewer: ```suggestion
In TypeQL, use `select` to filter out unwanted data:
```

[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has name $name, has email $email;
select $name, $email;	# eliminates $p

> ----

Reviewer: Worth mentioning that our `select` is unordered unlike SQL.

> Special case: if you're returning data and want to format it through a `fetch` clause, it can also work as a projection:

Reviewer: You've already introduced `fetch` and `select` as equal options in the introductory sentence, this should not be a special case. Something like this maybe?
```suggestion
You can also use `fetch` to format the data in a JSON-like document:
```

[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has name $name, has email $email;
fetch {
  "name": $name,
  "email": $email
};

> ----

Reviewer: I'm tempted to include `$p.name` kind of projections here too. It might be too much new information in one place, but showing how concise we can be is probably good.

Author: good idea!

=== WHERE → Pattern Constraints

SQL's `WHERE` clause filters rows. In TypeQL, constraints are part of the `match` pattern.

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE age > 25;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=2]
match
  $p isa person, has name $name, has age $age;
  $age > 25;
select $name;
----

Shorthand:
[,typeql]
----
#!test[read, count=2]
match
  $p isa person, has name $name, has age > 25;
select $name;
----

==== AND/OR Conditions in WHERE clauses

> SQL uses `AND` and `OR` in `WHERE` clauses. TypeQL uses pattern conjunction (comma or semicolon) and disjunction (`or`).

Reviewer: Awkward phrasing. I'd suggest:
```
In TypeQL, patterns delimited by a semicolon, as well as constraints within a pattern, are in a conjunction (i.e. implicitly `and`ed); use an `or` pattern to represent a disjunction.
```
This could still use more work.

> **SQL:**

Reviewer: `**AND in SQL:**`, `**Conjunction in TypeQL:**` etc? For consistency.

FROM person p
RIGHT JOIN employment e ON p.id = e.employee_id;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=1]
match
  $e isa employment;
  try {
    $p isa person, has name $pname;
    $e links (employee: $p);
  };
select $pname;
----

This returns all employments, with employee names if they exist.
Employments without matching persons will still appear in the results (with `$pname` unbound).

==== CROSS JOIN

A CROSS JOIN returns the Cartesian product of two tables - every row from the first table combined with every row from the second table.
In TypeQL, match entities from both types independently without requiring any relation between them.

**SQL:**
[source,sql]
----
SELECT p.name, e.id
FROM person p
CROSS JOIN employment e;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has name $pname;
  $e isa employment;
select $pname, $e;
----

This returns every combination of person and employment, regardless of whether they have any relationship.
The result set size is the product of the number of persons and employments.

==== FULL OUTER JOIN

A FULL OUTER JOIN returns all rows from both tables, with NULL values where there's no match.

> TypeQL doesn't yet have an efficient equivalent of full outer joins! Please leave us a message if this operation is required for your application.

Reviewer: This reads as a bit defeatist right now. The first sentence could be closer to "Full joins do not translate well to TypeQL. If your application requires full joins, ..."

Author: fair haha

----
SELECT 
  COUNT(*) as total,
  AVG(age) as avg_age,
  MAX(age) as max_age,
  MIN(age) as min_age,
  SUM(age) as sum_age
FROM person;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=1]
match
  $p isa person, has age $age;
reduce
  $total = count,
  $avg_age = mean($age),
  $max_age = max($age),
  $min_age = min($age),
  $sum_age = sum($age);
----

=== HAVING → reduce + match

**SQL:**
[source,sql]
----
SELECT age, COUNT(*) as count
FROM person
GROUP BY age
HAVING COUNT(*) > 1;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=0]
match
  $p isa person, has age $age;
reduce $count = count groupby $age;
match $count > 1;
----

== Sorting and Pagination

=== ORDER BY → sort

**SQL:**
[source,sql]
----
SELECT name, age FROM person ORDER BY age DESC, name ASC;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has name $name, has age $age;
sort $age desc, $name asc;
select $name, $age;
----

=== LIMIT and OFFSET

**SQL:**
[source,sql]
----
SELECT name FROM person ORDER BY name LIMIT 10 OFFSET 20;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=0]
match
  $p isa person, has name $name;
sort $name;
offset 20;
limit 10;
select $name;
----

> 

Reviewer: I believe `offset` must come before `limit` here, worth a note with explanation of sequencing.

== Advanced Operations

=== DISTINCT → distinct

**SQL:**
[source,sql]
----
SELECT DISTINCT age FROM person;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has age $age;
select $age;
distinct;
----

> === Subqueries → Functions or Nested match

Reviewer: "Nested match" is not a thing we have

**SQL:**
[source,sql]
----
SELECT name FROM person
WHERE age > (SELECT AVG(age) FROM person);
----

> **TypeQL using functions:**

Reviewer: ```suggestion
**TypeQL:**
```

[,typeql]
----
#!test[read, count=1]
with fun avg_age() -> integer:
  match $p isa person, has age $age;
  return mean($age);
match
  $p isa person, has name $name, has age $age;
  $age > avg_age();
select $name;
----

=== UNION → or patterns

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE age < 25
UNION
SELECT name FROM person WHERE age > 35;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=1]
match
  $p isa person, has name $name, has age $age;
  { $age < 25; } or { $age > 35; };
select $name;

> ----

Reviewer: I believe this is also valid:
```
match
  {
    $p isa person, has name $name, has age < 25;
  } or {
    $p isa person, has name $name, has age > 35;
  }
select $name;
```
Might be worth mentioning? It's closer to the original SQL, so may be either to translate, both for humans and LLMs.

Author: Good point, thanks

> === EXISTS → not { not { ... } }

Reviewer: Oof. Sugar for `not not` would be nice to have, huh.

**SQL:**
[source,sql]
----
SELECT name FROM person p
WHERE EXISTS (
  SELECT 1 FROM employment e
  WHERE e.employee_id = p.id OR e.employer_id = p.id
);
----

**TypeQL:**
[,typeql]
----
#!test[read, count=2]
match
  $p isa person, has name $name;

>   not { not { $e isa employment, links (employee: $p); }; };

Reviewer: This warrants a sentence explaining smth like "we're looking for a person such that it's _not_ true that there is _not_ an employment involving that person"

Reviewer: Maybe even in this code block with the comments breaking it down:
```php
not { # it is not true ...
not { # ... that there is no employment ...
 ... # ... such that ...
```

select $name;
----

> _This will be improved in future versions of TypeQL!_

Reviewer: Ha, glad we're in agreement.

=== IN → Multiple value constraints

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE age IN (25, 30, 35);
----

**TypeQL:**
[,typeql]
----
#!test[read, count=1]
match
  $p isa person, has name $name, has age $age;

>   { $age == 25; } or { $age == 30; } or { $age == 35; };

Reviewer: Aside: once we have lists, we should be able to say `$age in [25, 30, 35]`. Note that `$age` is an attribute while the list contains raw value, we should handle that.

Author: agree, we should make that work though I think

select $name;
----

=== LIKE → like (regex)

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE email LIKE '%@example.com';
----

**TypeQL:**
[,typeql]
----
#!test[read, count=3]
match
  $p isa person, has name $name, has email $email;
  $email like ".*@example.com";
select $name;
----

=== NULL checks → not { ... }

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE email IS NULL;
----

**TypeQL:**

> TypeQL doesn't have NULLs. A value is either existing or absent in the database:

Reviewer: "TypeDB doesn't _store_ NULLs" is more honest imo.

Author: fair

[,typeql]
----
#!test[read, count=0]
match
  $p isa person, has name $name;
  not { $p has email; };
select $name;
----

**SQL:**
[source,sql]
----
SELECT name FROM person WHERE email IS NOT NULL;
----

**TypeQL:**
[,typeql]
----
#!test[read, count=3]
match

>   $p isa person, has name $name, has email $email;

Reviewer: `has email;` (no variable) is closer to `WHERE email IS NOT NULL` in spirit.

select $name;
----

Note that for optional matches like when using LEFT/RIGHT joins, TypeQL's `try` construct can create `None` values. Unlike `null`s in SQL, a `None` is not a value that can be stored. It's also only possible to generated for explicitly known optional variables (ie. ones generated in `try` patterns.)  

> Fundamentally, in SQL a `null` is an extra value that any column could contain. In TypeQL, a `None` is a marker given to a variable when it is unset, but it is never used as a value in the data. In other words, any typical value in SQL has the range of possible values of: `<values> + null`. In TypeQL, variables that are explicitly optional can hold `Either<None, Value>` instead of just `Value`.

Reviewer: I don't think we have "explicitly optional variables" anymore. Variables are optional implicitly if they can only be bound in a `try` block.

----

=== Relations vs Foreign Keys

In SQL, relationships are represented through foreign keys. In TypeQL, relationships are explicit relations that can have their own attributes and connect multiple entities.

**SQL (with foreign keys):**
[source,sql]
----
CREATE TABLE employment (
  id INT PRIMARY KEY,
  employee_id INT REFERENCES person(id),
  employer_id INT REFERENCES company(id),
  start_date DATE
);
----

**TypeQL:**
[,typeql]
----
define
  relation employment, relates employee, relates employer, owns start-date;
  attribute start-date, value datetime;
  person plays employment:employee;
  company plays employment:employer;
----

== Summary Table

[cols="^1,^1,^1", options="header"]
|===
| SQL Construct | TypeQL Equivalent | Notes

| `SELECT` | `match` + `fetch`/`select` | Pattern matching replaces table selection
| `FROM` | Implicit in `match` patterns | Types are specified in the pattern
| `WHERE` | Constraints in `match` | Filtering is part of the pattern
| `JOIN` | Relation patterns | Relations naturally connect entities
| `INNER JOIN` | Required relation patterns | Both sides must exist
| `LEFT JOIN` | Optional patterns with `try` | Use `try` for optional matches

> | `RIGHT JOIN` | Optional patterns with `try` | Reverse pattern direction

Reviewer: This is missing `CROSS JOIN`

| `INSERT` | `insert` | Similar syntax, pattern-based
| `UPDATE` | `update` | Only for single-value attributes
| `DELETE` | `delete` | Pattern-based deletion
| `GROUP BY` | `reduce ... groupby` | Aggregation with grouping
| `HAVING` | `reduce` + `match` | Filter aggregated results
| `ORDER BY` | `sort` | Similar functionality
| `LIMIT` | `limit` | Identical
| `OFFSET` | `offset` | Identical
| `DISTINCT` | `distinct` | Identical

> | `COUNT/SUM/AVG/etc` | `reduce` with functions | `count`, `sum`, `mean`, etc.

Reviewer: ```suggestion
| `COUNT`/`SUM`/`AVG`/`etc. | `reduce` with `count`/`sum`/`mean`/etc. | Similar, assigned to a variable
```
